mlflow/projects/__init__.py

```diff
@@ -0,0 +1,282 @@
+"""APIs for running MLflow Projects locally or remotely."""
+
+from __future__ import print_function
+
+import hashlib
+import json
+import os
+import re
+import shutil
+import tempfile
+
+from distutils import dir_util
+
+
+from mlflow.projects._project_spec import Project
+from mlflow.entities.run_status import RunStatus
+from mlflow.entities.source_type import SourceType
+from mlflow.entities.param import Param
+import mlflow.tracking as tracking
+from mlflow.projects.submitted_run import SubmittedRun
+
+
+from mlflow.utils import file_utils, process
+from mlflow.utils.logging_utils import eprint
+
+# TODO: this should be restricted to just Git repos and not S3 and stuff like that
+_GIT_URI_REGEX = re.compile(r"^[^/]*:")
+
+
+class ExecutionException(Exception):
+    """Exception thrown when executing a project fails."""
+    pass
+
+
+def _run(uri, entry_point="main", version=None, parameters=None, experiment_id=None,
+         mode=None, cluster_spec=None, git_username=None, git_password=None, use_conda=True,
+         use_temp_cwd=False, storage_dir=None, block=True):
+    exp_id = experiment_id or tracking._get_experiment_id()
+    if mode is None or mode == "local":
+        return _run_local(
+            uri=uri, entry_point=entry_point, version=version, parameters=parameters,
+            experiment_id=exp_id, use_conda=use_conda, use_temp_cwd=use_temp_cwd,
+            storage_dir=storage_dir, git_username=git_username, git_password=git_password,
+            block=block)
+    if mode == "databricks":
+        from mlflow.projects.databricks import run_databricks
+        return run_databricks(
+            uri=uri, entry_point=entry_point, version=version, parameters=parameters,
+            experiment_id=exp_id, cluster_spec=cluster_spec, git_username=git_username,
+            git_password=git_password)
+    supported_modes = ["local", "databricks"]
+    raise ExecutionException("Got unsupported execution mode %s. Supported "
+                             "values: %s" % (mode, supported_modes))
+
+
+def run(uri, entry_point="main", version=None, parameters=None, experiment_id=None,
+        mode=None, cluster_spec=None, git_username=None, git_password=None, use_conda=True,
+        use_temp_cwd=False, storage_dir=None, block=True):
+    """
+    Run an MLflow project from the given URI.
+
+    Supports downloading projects from Git URIs with a specified version, or copying them from
```

Comment:
We should mention that this throws an ExecutionException if it fails in blocking mode.